src/processors/tone/ladder_filter/LadderFilterProcessor.cpp

```diff
@@ -0,0 +1,68 @@
+#include "../../ParameterHelpers.h"
+
+#include "LadderFilterProcessor.h"
+
+LadderFilterProcessor::LadderFilterProcessor (UndoManager* um) : BaseProcessor ("Ladder Filter", LadderParameters::createParameterLayout(), um), p (vts)
+{
+    // Info and style
+    uiOptions.info.description = "Resonant 4-pole ladder high-pass and low-pass,\n capable of self-oscillation. With nonlinear drive.";
+    uiOptions.info.authors = StringArray { "Butch Warns" };
+    uiOptions.info.infoLink = "https://butchwarns.de/";
+    uiOptions.backgroundColour = juce::Colour::fromRGB (0xBB, 0xE1, 0xC3);
+    uiOptions.powerColour = juce::Colour::fromRGB (0x99, 0x0B, 0x3C);
+}
+
+void LadderFilterProcessor::prepare (double sampleRate, int samplesPerBlock)
+{
+    // Initialize filters
+    for (int channel = 0; channel < 2; ++channel)
+    {
+        lp[channel].reset (sampleRate);
+        hp[channel].reset (sampleRate);
+    }
+
+    // Initialize parameter smoothing
+    p.reset (sampleRate, samplesPerBlock);
+}
+
+void LadderFilterProcessor::processAudio (AudioBuffer<float>& buffer)
+{
+    // Get parameter values for current block
+
+    const double hp_cutoff = p.hp_cutoff();
+    const double hp_resonance = p.hp_resonance();
+    const double lp_cutoff = p.lp_cutoff();
+    const double lp_resonance = p.lp_resonance();
+
+    const double gain = p.drive();
+
+    // Update filter parameters
+    for (int channel = 0; channel < 2; ++channel)
+    {
+        hp[channel].set_cutoff (hp_cutoff);
+        hp[channel].set_resonance (hp_resonance);
+        lp[channel].set_cutoff (lp_cutoff);
+        lp[channel].set_resonance (lp_resonance);
+    }
+
+    // Iterate over frames
+    for (int n = 0; n < buffer.getNumSamples(); ++n)
+    {
+        // Process samples in the current frame
+        for (int channel = 0; channel < buffer.getNumChannels(); ++channel)
+        {
+            // Reference the input sample for cleaner calculations
+            float& x = buffer.getWritePointer (channel)[n];
+
+            //Apply filters
+            x = hp[channel].process (x);
+            x = lp[channel].process (x);
+
+            // Appply drive
+            x *= gain;
+
+            // Saturate (static nonlinearity) and blend using drive param
+            x = p.drive_normalized() * fast_tanh_2 (x) + (1.0 - p.drive_normalized()) * x;
+        }
+    }
```

Comment:
So I've been noticing a bit of "crackling" when moving the sliders around, and I think that's coming from the fact the the parameter smoothing seems to be happening per-block rather than per-sample.

I think what I'd recommend is to use the `chowdsp::SmoothedBufferValue` for smoothing the parameters (the new [Flapjack](https://github.com/Chowdhury-DSP/BYOD/blob/main/src/processors/drive/flapjack/Flapjack.h#L18-L20) processor has an example of this). That way we can know when the parameter needs to be smoothed, and when it's staying constant. `SmoothedBufferValue` also has a way of setting a "mapping function" which can be used to apply "skews" to the parameters when needed.

From there, we can re-write this top-level processing block in a way that should help both the smoothing, and the CPU performance:
```cpp
const auto num_samples = buffer.getNumSamples();

// Set up parameter smoothers
hp_cutoff_smooth.process (numSamples); // hp_cutoff_smooth is a chowdsp::SmoothedBufferValue
hp_resonance_smooth.process (numSamples);
// do the rest of the parameters...

for (int channel = 0; channel < buffer.getNumChannels(); ++channel)
{
    auto* data = buffer.getWritePointer (channel);

    // process hpf
    if (hp_cutoff_smooth.isSmoothing() || hp_resonance_smooth.isSmoothing())
    {
        // if we need parameter smoothing...
        const auto* cutoff_smooth = hp_cutoff_smooth.getSmoothedBuffer();
        const auto* resonance_smooth = hp_resonance_smooth.getSmoothedBuffer();
        for (int n = 0; n < num_samples; ++n)
        {
            hp[channel].set_cutoff (cutoff_smooth[n]);
            hp[channel].set_resonance (resonance_smooth[n]);
            data[n] = hp[channel].process (data[n]);
        }
    }
    else
    {
        // no need for parameter smoothing
       hp[channel].set_cutoff (hp_cutoff_smooth.getCurrentValue());
       hp[channel].set_resonance (hp_resonance_smooth.getCurrentValue());
        for (int n = 0; n < num_samples; ++n)
        {
            data[n] = hp[channel].process (data[n]);
        }
    }

    // here we would do the same thing for the lowpass filter

    // apply drive
    juce::FloatVectorOperations::multiply (data, drive.getSmoothedBuffer(), num_samples);

    // saturate and blend
    const auto* drive_norm_smooth = drive_normalize_smooth.getSmoothedBuffer();
    for (int n = 0; n < num_samples; ++n)
        data[n] = drive_norm_smooth[n] * fast_tanh (data[n]) + (1.0f - drive_norm_smooth[n]) * data[n];
}
```

Anyway, that's a rough outline, but let me know if you have any questions or other ideas for how to do this sort of thing.

Reply:
Good point! I also think the smoothing can be improved. I will try to implement this as per your example, or the Flapjack processor respectively.

Reply:
Hi Jatin,

I was able to implement what you suggested by using smoothedBufferValues in the processor. The fork is now updated for you to have a look, if you feel like it! Everything seems to work, just the safety limit on resonance (popup parameter) is now no longer implemented. I am still contemplating if there is a way of doing this without excessive per-sample if statements in the new processor. Do you by any chance have an idea?

-Butch